probe_src/probe_frontend/src/arena.rs

```diff
@@ -0,0 +1,370 @@
+
+        let count = (header.used - size_of::<ArenaHeader>()) / size_of::<ffi::Op>();
+
+        log::debug!("[unsafe] converting Vec<u8> to &[ffi::Op]");
+        let ops = unsafe {
+            let ptr = bytes.as_ptr().add(size_of::<ArenaHeader>()) as *const ffi::Op;
+            std::slice::from_raw_parts(ptr, count)
+        };
+
+        Ok(Self { raw: bytes, ops })
+    }
+
+    pub fn decode(self, ctx: &ArenaContext) -> Result<Vec<ops::Op>> {
+        self.ops
+            .iter()
+            .map(|x| ops::Op::ffi_from(x, ctx))
+            .collect::<Result<Vec<_>>>()
+            .wrap_err("Failed to decode arena ops")
+    }
+}
+
+/// This struct represents a single `data/*.dat` arena allocator file emitted by libprobe.
+pub struct DataArena {
+    header: ArenaHeader,
+    raw: Vec<u8>,
+}
+
+impl DataArena {
+    pub fn from_bytes(bytes: Vec<u8>) -> Result<Self> {
+        if bytes.len() < size_of::<ArenaHeader>() {
+            return Err(eyre!(
+                "Arena buffer too small, got {}, minimum size {}",
+                bytes.len(),
+                size_of::<ArenaHeader>()
+            ));
+        }
+        let header = unsafe { get_header_unchecked(&bytes) };
+        if header.capacity != bytes.len() {
+            return Err(eyre!(
+                "Invalid arena capacity, expected {}, got {}",
+                header.capacity,
+                bytes.len(),
+            ));
+        }
+        if header.used > header.capacity {
+            return Err(eyre!(
+                "Arena size {} is greater than capacity {}",
+                header.used,
+                header.capacity,
+            ));
+        }
+        Ok(Self { header, raw: bytes })
+    }
+
+    pub fn try_deref(&self, ptr: usize) -> Option<*const u8> {
+        match ptr >= self.header.base_address
+            && ptr <= (self.header.base_address + self.header.used)
+        {
+            false => None,
+            true => Some(unsafe { self.raw.as_ptr().add(ptr - self.header.base_address) }),
+        }
+    }
+}
+
+/// this struct represents a `<TID>/data` directory from libprobe.
+pub struct ArenaContext(pub Vec<DataArena>);
+
+impl ArenaContext {
+    pub fn try_deref(&self, ptr: usize) -> Option<*const u8> {
+        for vec in self.0.iter() {
+            if let Some(x) = vec.try_deref(ptr) {
+                return Some(x);
+            }
+        }
+        None
+    }
+}
+
+/// Parse the front of a raw byte buffer into a libprobe arena header
+///
+/// # Safety:
+/// Invoking this function on any byte buffer smaller than [`std::mem::size_of<ArenaHeader>()`]
+/// bytes is undefined behavior (best case a segfault). Invoking this method on a byte buffer
+/// that's not a valid libprobe arena will produce garbage values that should not be used.
+unsafe fn get_header_unchecked(bytes: &[u8]) -> ArenaHeader {
+    let ptr = bytes as *const [u8] as *const ArenaHeader;
+    log::debug!("[unsafe] converting byte buffer into ArenaHeader");
```

Comment:
Let's include the 4 values in the log (with base address formatted as a pointer). If something crashes later on, that would really help us spot the invalid header. We can also assert that `used <= capacity`. We could even pass in the `int(fname.strip_ext())` as an arg and assert that it equals the `instantiation` number.

Reply:
That's a good idea: I had a check for `used <= capacity` in both the places I called this function, but for code deduplication if nothing else it makes sense to move it inside the function. 

As a side-point, generally I prefer to return an `Err` rather than fail an assertion and panic even for non-recoverable errors because then I get the error trace from `color-eyre`. Currently, all those assertions are simple memory arithmetic comparisons, which—to me—make sense to run all the time, if more complicated assertion-like-statements are introduced we can get the best of both worlds by using `if cfg!(debug_assertions) {...}`

Reply:
Done in https://github.com/charmoniumQ/PROBE/pull/20/commits/70649ba0efd15a8ae10e0f257aeadf41a866fb43.